### PR TITLE
Feature: Disconnect account

### LINKED ISSUE #2005: [settings] "Disconnect" functionality
## Description

The "Disconnect" button deletes the account keys stored in the Stripe settings.

Screen Shot 2021-09-29 at 6 59 07 PM

(use `wp/data` implementation from https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2004 ).

Show a confirmation dialog upon clicking.
Screen Shot 2021-09-30 at 8 32 48 AM

Please consult with UX on functionality. In particular:
- Whether this erases both test and live account keys/hooks
- Whether the page should be refreshed upon confirmation, to show the onboarding again (I'm assuming so, but better confirm)

### PR DESCRIPTION
**Pull request by Mayisha**
Fixes #2005

### Description
The `Disconnect` button in the `Account Details` section shows a confirmation modal. After confirming to disconnect via modal all the live and test account keys are removed from DB. 

diconnect

### Testing instructions
- Make sure you have test keys or live keys set
- Enable `_wcstripe_feature_upe_settings`
- Go to the `Settings` tab in `http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe`
- Click `Disconnect` from the 3 dot menu of the `Account Details` section and the modal will appear

### QA steps
- [ ]  Click the `Cancel` or `Cross` icon on the modal -> Modal disappears and nothing changes, previous keys are still present in DB
- [ ] Click `Disconnect` on the modal -> Modal disappears, the window reloads the `http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe` page and the "connect stripe account" card is now there on the page.
- [ ] All the test and live keys are empty in the DB

### CODE REVIEW

File: client/data/account-keys/actions.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ export function updateIsSavingAccountKeys( isSaving, error ) {
 	};
 }
 
-export function* saveAccountKeys() {
+export function* saveAccountKeys( disconnecting = false ) {
```
Comment (tommyshellberg) [inline client/data/account-keys/actions.js]: In my PR for #1835 I will be changing this function to accept an object representing the account keys. Perhaps rather than adding a parameter strictly for `disconnecting` we could allow this to take an object for keys and check if the keys are empty?

Comment (tommyshellberg) [reply client/data/account-keys/actions.js]: The PR isn't finished but I pushed up the draft branch, you can see here what I did to saveAccountKeys():

https://github.com/woocommerce/woocommerce-gateway-stripe/compare/feat/account-keys-ui?expand=1#diff-512bec88a04a8f9f319204ee46deefab02432fda511ef85ff689968f0e63eeadR29

It no longer reads from the Redux store but rather gets the keys to save directly as an argument. When the save works without an error, _then_ the Redux store is saved.
This was mainly because we didn't want to update the Redux store before the keys were actually saved in case the save fails or someone closes the modal without saving(then the Redux store would persist that information).

Do you think that could work for you?

Otherwise, maybe we could just create a separate function like clearAccountKeys().

**Comment by tommyshellberg** [comment]
This works well and looks good on mobile. I'm wondering if we could make the UX a bit better by utilizing the `isBusy` prop of the Disconnect button and set it to `true` while the `saveAccountKeys()` function is working. 

While the function is running it will set `isSaving` to `true` so you can listen to that value for the `isBusy` prop.
Additionally, in order to actually show that, the modal would have to not instantly close. If `saveAccountKeys()` was used we could check for it being done before running the `onClose()` function. In my testing I found that I could just `await` the call to `saveAccountKeys` but it's technically a generator function and not an async one. So I'm not sure why that works 😆 maybe you have a better idea?

File: client/settings/payment-settings/disconnect-stripe-confirmation-modal.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		onClose();
+		updateAccountKeys( {
+			publishable_key: '',
+			secret_key: '',
+			webhook_secret: '',
+			test_publishable_key: '',
+			test_secret_key: '',
+			test_webhook_secret: '',
+		} );
+		saveAccountKeys( true ).then( () => {
```
Comment (tommyshellberg) [inline client/settings/payment-settings/disconnect-stripe-confirmation-modal.js]: I came across this bug myself when testing for #1835; when the keys are cleared the app doesn't update to reflect that the onboarding component should be loaded. I wonder if we should just treat that as a bug and fix it to re-render properly rather than patch it with a page reload? 🤔

File: client/data/account-keys/actions.js (hunk 1 above)
Comment (Mayisha) [reply client/data/account-keys/actions.js]: > In my PR for #1835 I will be changing this function to accept an object representing the account keys. Perhaps rather than adding a parameter strictly for disconnecting we could allow this to take an object for keys and check if the keys are empty?

I have updated my PR following this. Thank you for pointing me to the related issue :100:

File: client/settings/payment-settings/disconnect-stripe-confirmation-modal.js (hunk 2 above)
Comment (Mayisha) [reply client/settings/payment-settings/disconnect-stripe-confirmation-modal.js]: I have removed the patch. As the same issue happens when adding the keys as well, we can handle the re-render related issue separately. :+1: 

Have you already created an issue for that @tommyshellberg ?

**Comment by Mayisha** [comment]
@tommyshellberg I have updated the PR and addressed your feedback in the [latest commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2056/commits/e15730e4a5cff6b0f696ea62527eb3c2f6b0e594). Please review this again.

> Additionally, in order to actually show that, the modal would have to not instantly close. If saveAccountKeys() was used we could check for it being done before running the onClose() function

I actually realized that explicitly calling the `onClose` is not required at all as it disappears automatically when the keys are cleared and it's parent component gets unmounted.

File: client/data/account-keys/actions.js [hunk 3]
```diff
@@ -26,11 +26,11 @@ export function updateIsSavingAccountKeys( isSaving, error ) {
 	};
 }
 
-export function* saveAccountKeys() {
+export function* saveAccountKeys( accountKeys ) {
+	const isDisconnecting =
+		! accountKeys.publishable_key && ! accountKeys.test_publishable_key;
```
Comment (tommyshellberg) [inline client/data/account-keys/actions.js:31]: This is working, thanks for updating to use `saveAccountKeys()`. There is an edge case with this implementation in which someone could save the form with the publishable keys empty(test and live) but technically have everything else filled out and prompt a disconnect. Is it worth worrying about or is it unlikely enough to happen to 🚢 ?

**Review by tommyshellberg — APPROVED** [review]

File: client/settings/payment-settings/disconnect-stripe-confirmation-modal.js (hunk 2 above)
Comment (tommyshellberg) [reply client/settings/payment-settings/disconnect-stripe-confirmation-modal.js]: Created here:

https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2075

Let me know if the diagnosis makes sense. 🤓

File: client/data/account-keys/actions.js (hunk 3 above)
Comment (Mayisha) [reply client/data/account-keys/actions.js:31]: From a discussion on [another PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2024#discussion_r722259720), I have this impression that in the new settings all the 3 keys are required for the corresponding mode. That's why I thought of checking only these two keys.

We can check if all the keys are empty as well. What do you think?
